File: src/Landbridge.Classifier/destroy-guard.mjs
/**
 * Qwen Auto Mode Layer 0 destroy guard (Apache-2.0, QwenLM/qwen-code
 * packages/core/src/permissions/destructive-commands.ts, b4f5079).
 *
 * Qwen hard-blocks these unless the last user prompt clearly asked to discard
 * or destroy. Landbridge never AutoDenies: a match is Ask, and we do not take
 * a user-prompt override (the classify request does not carry that text).
 */

const DESTRUCTIVE_GIT = Object.freeze([
  /\bgit\s+reset\s+--hard\b/,
  /\bgit\s+checkout\s+--\s+\./,
  /\bgit\s+clean\s+-[a-zA-Z]*f/,
  /\bgit\s+stash\s+drop\b/,
]);

const GIT_AMEND = /\bgit\s+commit\s+--amend\b/;

const IAC_DESTROY = Object.freeze([
  /\bterraform\s+destroy\b/,
  /\bpulumi\s+destroy\b/,
  /\bcdk\s+destroy\b/,
]);

function stripShellQuotes(command) {
  return command.replace(
    /(?:^|\s)(?:bash|sh|zsh|fish|dash|ksh)\s+-[a-zA-Z]*c\s+(?:"([^"]*)"|'([^']*)')/g,
    (_match, dq, sq) => " " + (dq ?? sq ?? ""),
  );
}

/**
 * @returns {{ blocked: true, reason: string } | null}
 */
export function matchDestructiveCommand(command) {
  if (typeof command !== "string" || !command.trim()) return null;
  const expanded = command + " " + stripShellQuotes(command);

  for (const pattern of DESTRUCTIVE_GIT) {
    if (pattern.test(expanded)) {
      const matched = command.match(pattern)?.[0] ?? command;
      return {
        blocked: true,
        reason: `Destructive git command: "${matched}".`,
      };
    }
  }

  if (GIT_AMEND.test(expanded)) {
    return {
      blocked: true,
      reason: 'Blocked "git commit --amend" (not known to be this session\'s commit).',
    };
  }

  for (const pattern of IAC_DESTROY) {
    if (pattern.test(expanded)) {
      const toolName = command.match(pattern)?.[0]?.split(/\s+/)[0] ?? "unknown";
      return {
        blocked: true,
        reason: `Infrastructure destroy command: "${toolName} destroy".`,
      };
    }
  }

  return null;
}

File: src/Landbridge.Classifier/DestroyGuard.cs
using System.Text.RegularExpressions;

namespace Landbridge.Classifier;

/// <summary>
/// Ask (never Deny) on a small list of discard/destroy commands. A match skips
/// the LLM so a model outage cannot wave them through.
/// </summary>
public static class DestroyGuard
{
    private static readonly Regex[] DestructiveGit =
    [
        new(@"\bgit\s+reset\s+--hard\b", RegexOptions.CultureInvariant | RegexOptions.Compiled),
        new(@"\bgit\s+checkout\s+--\s+\.", RegexOptions.CultureInvariant | RegexOptions.Compiled),
        new(@"\bgit\s+clean\s+-[a-zA-Z]*f", RegexOptions.CultureInvariant | RegexOptions.Compiled),
        new(@"\bgit\s+stash\s+drop\b", RegexOptions.CultureInvariant | RegexOptions.Compiled),
    ];

    private static readonly Regex GitAmend =
        new(@"\bgit\s+commit\s+--amend\b", RegexOptions.CultureInvariant | RegexOptions.Compiled);

    private static readonly Regex[] IacDestroy =
    [
        new(@"\bterraform\s+destroy\b", RegexOptions.CultureInvariant | RegexOptions.Compiled),
        new(@"\bpulumi\s+destroy\b", RegexOptions.CultureInvariant | RegexOptions.Compiled),
        new(@"\bcdk\s+destroy\b", RegexOptions.CultureInvariant | RegexOptions.Compiled),
    ];

    private static readonly Regex ShellC =
        new(@"(?:^|\s)(?:bash|sh|zsh|fish|dash|ksh)\s+-[a-zA-Z]*c\s+(?:""([^""]*)""|'([^']*)')",
            RegexOptions.CultureInvariant | RegexOptions.Compiled);

    public static (bool Blocked, string Reason) Match(string? command)
    {
        if (string.IsNullOrWhiteSpace(command))
            return (false, "");

        var expanded = command + " " + ShellC.Replace(command, m =>
            " " + (m.Groups[1].Success ? m.Groups[1].Value : m.Groups[2].Value));

        foreach (var pattern in DestructiveGit)
        {
            var m = pattern.Match(expanded);
            if (m.Success)
                return (true, $"Destructive git command: \"{m.Value}\".");
        }

        if (GitAmend.IsMatch(expanded))
            return (true, "Blocked \"git commit --amend\" (not known to be this session's commit).");

        foreach (var pattern in IacDestroy)
        {
            if (pattern.IsMatch(expanded))
            {
                var tool = pattern.Match(command).Value.Split(' ', 2)[0];
                return (true, $"Infrastructure destroy command: \"{tool} destroy\".");
            }
        }

        return (false, "");
    }
}
